nl_server/config_reader.py

```diff
@@ -0,0 +1,214 @@
+    all_paths.append(_CATALOG_MOUNT_PATH)
+
+  # Then check user set config file
+  # TODO: make this generic and independent of custom DC.
+  # Currently this depends on the assumption of custom DC user sub path.
+  user_data_path = custom_dc_util.get_custom_dc_user_data_path()
+
+  if gcs.is_gcs_path(user_data_path):
+    full_gcs_path = os.path.join(user_data_path, _CATALOG_USER_PATH_SUFFIX)
+    if gcs.download_blob_by_path(full_gcs_path, _CATALOG_TMP_PATH):
+      all_paths.append(_CATALOG_TMP_PATH)
+  else:
+    full_user_path = os.path.join(user_data_path, _CATALOG_USER_PATH_SUFFIX)
+    if os.path.exists(full_user_path):
+      all_paths.append(full_user_path)
+
+  # Now load and merge all the catalog config files
+  catalog = Catalog(
+      version='',
+      indexes={},
+      models={},
+  )
+  for p in all_paths:
+    logging.info('Loading index and model catalog from: %s', p)
+    with open(p) as f:
+      partial_catalog = yaml.safe_load(f.read())
+      # read version
+      ver = catalog.version
+      this_ver = partial_catalog['version']
+      if (ver and this_ver and ver != this_ver):
+        raise ValueError(
+            f'Inconsistent version in config files: {ver} and {this_ver}')
+
+      if this_ver not in _SUPPORTED_VERSIONS:
+        raise ValueError(f'Unknown version: {this_ver}')
+      catalog.version = this_ver
+
+      indexes = catalog.indexes
+      models = catalog.models
+
+      # TODO: when another version is added, extract separate version read as
+      # a function.
+      if this_ver == 1:
+        # read indexes
+        for index_name, index_config in partial_catalog['indexes'].items():
+          store_type = index_config['store_type']
+          match store_type:
+            case StoreType.MEMORY:
+              indexes[index_name] = MemoryIndexConfig(**index_config)
+            case StoreType.LANCEDB:
+              indexes[index_name] = LanceDBIndexConfig(**index_config)
+            case StoreType.VERTEXAI:
+              indexes[index_name] = VertexAIIndexConfig(**index_config)
+            case _:
+              raise ValueError(f'Unknown store type: {store_type}')
+
+          # read models
+          for model_name, model_config in partial_catalog['models'].items():
+            model_type = model_config['type']
+            match model_type:
+              case ModelType.LOCAL:
+                models[model_name] = LocalModelConfig(**model_config)
+              case ModelType.VERTEXAI:
+                models[model_name] = VertexAIModelConfig(**model_config)
+              case _:
+                raise ValueError(f'Unknown model type: {model_type}')
+  logging.debug(json.dumps(asdict(catalog), indent=2))
+  return catalog
+
+
+def read_env() -> Env:
+  """Reads the envs
+  """
+  # TODO: Make this generic by checking a user provided config file path
+  if custom_dc_util.is_custom_dc():
+    c = from_dict(data_class=Env, data=yaml.safe_load(open(_ENV_USER_PATH)))
```

Comment:
Hmm - this "leaks" the fact that `Env` is a `dataclass`. How about adding a class level `from_dict` method to `Env` itself and using `dacite` in there? It will keep the dataclass aspect encapsulated in the class and will be more flexible.